### PR TITLE
pull request for issue #65

### LINKED ISSUE #65: Cache type not usable if global sys_tmp_dir not writeable
### BC Break Report

|    Q        |   A
|------------ | ------
| Version     | 2.3.x-dev

#### Summary
When you define an own cache_dir, the system always tries the default sys_get_temp_dir first, even though you are not using it. If anything is wrong with it (like no write access), you cannot create a FileSystem Adapter at all.

#### Previous behavior

The sys_get_temp_dir() was only used in case that no other dir was defined.

#### Current behavior

I've provided the following config for the filesystem:

```
'name'    => 'filesystem',
'adapter' => 'filesystem',
'options' => [
    'cache_dir' => 'data/cache/',
    'ttl'       => '3600'
],
'plugins' => [
    [
        'name'    => 'exception_handler',
        'options' => [
            'throw_exceptions' => true,
        ],
    ],
    [
        'name' => 'Serializer'
    ]
]
```

As you can see, the **cache_dir** is provided in the settings. 
However, since the new version, the FileSystemOptions always calls the setCacheDir with the parameter `null`:
 `$this->setCacheDir(null);` (`FileSystemOptions.php:115`)

This results in a fallback to sys_get_temp_dir(), which will then be checked and - on my test maschine - was not writable. The Adapter throws and cannot be created then, though right after that, it would overwrite the CacheDir with the correct and working path.

#### How to reproduce
Define your `sys_temp_dir` in your `php.ini` to a non-existing or not-writable path. 
Then try to create an adapter with a valid `cache_dir` configuration. 
It will not work and fails with one exception of the method `normalizeCacheDirectory`

### PR DESCRIPTION
**Pull request by tck**

### CODE REVIEW

File: src/FilesystemOptions.php [hunk 1]
```diff
@@ -112,9 +112,11 @@ public function __construct($options = null)
             $this->dirPermission  = false;
         }
 
-        $this->setCacheDir(null);
-
         parent::__construct($options);
+
+        if (null === $this->cacheDir) {
+            $this->setCacheDir(null);
+        }
```
Comment (Ocramius) [inline src/FilesystemOptions.php:119]: Needs some sort of test / reproducer, IMO.

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: src/FilesystemOptions.php (hunk 1 above)
Comment (boesing) [reply src/FilesystemOptions.php:119]: I would suggest creating a directory in that test, change permissions to read-only using `chmod` and pass that to `ini_set` to modify sys temp dir for that test. 

Having something like `try {} finally {}` to restore temp dir back to default to ensure test does not introduce side-effects to other tests.

**Comment by sbenischke** [comment]
Any update on this topic?

In my opinion the `sys_tmp_dir` should be used as a fallback **after** the try to initialize the configured directory. So at least there is the possibility to get this adapter working, if the `sys_tmp_dir` is neither readable or writeable.

**Comment by boesing** [comment]
Once the requested changes are applied, we can move on.
Feel free to use these changes to create a new PR with added test(s) or wait until the author is responding.

File: src/FilesystemOptions.php (hunk 1 above)
Comment (bcremer) [reply src/FilesystemOptions.php:119]: `sys_temp_dir` is of type `PHP_INI_SYSTEM` and thus cannot be set during runtime via `ini_set`.

Setting the `TMPDIR`environment variable works:
`putenv('TMPDIR=/fasel')`.

Edit:
Seems like `putenv('TMPDIR=/fasel')` will also not work in the phpunit context.

**Comment by bcremer** [comment]
I stumbled about this problem during a `docker run  --read-only` deployment of an mezzio application.

I came up with the same fix as @tck. Building a reproducer as phpunit testcase is tricky because `sys_tmp_dir` cannot be modified during runtime, see: https://github.com/laminas/laminas-cache-storage-adapter-filesystem/pull/73#discussion_r1467308698.


A standalone reproducer looks like the following:

```php
<?php

require __DIR__  . '/vendor/autoload.php';

putenv('TMPDIR=/fasel');

// the following will throw an exception bacause `/fasel` does not exists even though the cache dir was explicitly overwritten
new \Laminas\Cache\Storage\Adapter\FilesystemOptions(['cacheDir' => '/tmp']);
```

```
PHP Fatal error:  Uncaught Laminas\Cache\Exception\InvalidArgumentException: Cache directory '/fasel' not found or not a directory in /home/apidev/current/laminas-cache-storage-adapter-filesystem/src/FilesystemOptions.php:440
```

If nobodoy has an idea how to codify the reproducer in phpunit I would suggest to move on and merge the PR as is.

**Comment by boesing** [comment]
Might be possible to modify the `tmpdir` setting via `php` CLI command when running a specific test group.

We could create a test, add a group to it, use that group in `additional_checks` when running `php vendor/bin/phpunit` along with the `-d` parameter to modify the temp dir target, add a check to that test that it skips the execution when tmp directory is readable and that should do the trick, no?

The check in the test could look like:

```php
if (is_writable(sys_get_temp_dir())) { 
    self::markAsRisky(); 
    self::markTestSkipped('Test has to be executed in read-only environment...');
}
```

---

Here is an example on how `addtional_checks` could look like:

https://github.com/laminas/laminas-cache/blob/ce069bc6c0c10abf13a51da1b0153090dc1caf0c/.laminas-ci.json#L4-L11

File: src/FilesystemOptions.php (hunk 1 above)
Comment (boesing) [reply src/FilesystemOptions.php:119]: I am not sure if this is a tmpdir only issue. Whenever a directory is used without write-permissions, that could lead to problems, no?

**Comment by bcremer** [comment]
@boesing Implemented in a separate PR: https://github.com/laminas/laminas-cache-storage-adapter-filesystem/pull/84

**Comment by boesing** [comment]
Superseded by #84

**Pull request closed without merging**